### DISCUSSION TITLE
how to add white space

### DISCUSSION THREAD

**Discussion by zrwsmd** [eclipse-langium/langium#1447] [Q&A] [answered]
my langium file is below
[demo.txt](https://github.com/eclipse-langium/langium/files/14926667/demo.txt)
how to add whitespace between 'IF' and 'THEN'

![image](https://github.com/eclipse-langium/langium/assets/26995040/3b3d8d28-bb9f-44fc-91fc-52cc7bc0ca3f)

**Comment by msujew** [comment]
Hey @zrwsmd,

you're probably looking for [our formatting guide](https://langium.org/guides/formatting/). Formatting is the preferred way of ensuring that most reasonable white space rules are satisfied.

**Reply by zrwsmd** [reply]
i mean when i validate a st file like 

IF state > 9THEN
   //operation
END_IF

it should report error
since this st file is written by user,so they can write maybe without whitespace

**Reply by msujew** [reply, marked as answer]
You can write a validation rule that ensures this based on the CST:

```ts
checkWhitespace(ifStatement: IfStatement, acceptor: ValidationAcceptor): void {
  const text = (getDocument(ifStatement).parseResult.value.$cstNode as RootCstNode).fullText;
  const thenKeyword = GrammarUtils.findNodeForKeyword(ifStatement.$cstNode, 'THEN');
  if (thenKeyword) {
    const index = thenKeyword.offset;
    const previousChar = text.charAt(index - 1);
    if (previousChar !== ' ') {
      acceptor('error', ...);
    }
  }
}
```

**Reply by zrwsmd** [reply]
for elsifConditions,it may have more than one condition like this,
![image](https://github.com/eclipse-langium/langium/assets/26995040/bc6fe728-6c0b-440b-8e60-b2f98c05e2e1)


if i write like this code,
![image](https://github.com/eclipse-langium/langium/assets/26995040/5430e125-33fb-40f1-9190-63ac23f0cdd1)
it  only validate first scrrenshot error part,can not  validate later 'THEN',even i change GrammarUtils.findNodeForKeyword third param to 4,5....

**Reply by msujew** [reply]
There is also a `findNodesForKeyword` function available that gives you all nodes.

**Reply by zrwsmd** [reply]
i have noticed this function,but seems did not work,only validate first 'THEN' appear position
![image](https://github.com/eclipse-langium/langium/assets/26995040/464d26bd-1f67-4e91-9671-3434550cd024)
![image](https://github.com/eclipse-langium/langium/assets/26995040/facaeedf-b115-4fd5-b077-e47e97bb19c0)


IF state > 9 THEN
    state := 1 / 8;
ELSIF state > 20 THEN
    state := 2;
ELSIF state > 50 THEN
    state := 5;
ELSIF state > 70 THEN
    state := 7;
ELSE
   state := 3;
END_IF;

**Reply by msujew** [reply]
Well you are applying it to the same property, so it will always position the error message at the same position. Use the `index` property to select at which index of the property the error is supposed to appear.

**Reply by zrwsmd** [reply]
so how to correct it, a little confused

```ts
  if (ifStatement) {
            const text = (AstUtils.getDocument(ifStatement).parseResult.value.$cstNode as RootCstNode).fullText;
            const thenKeywordArr = GrammarUtils.findNodesForKeyword(ifStatement.$cstNode, 'THEN');
            thenKeywordArr.forEach(thenKeyword => {
                if (thenKeyword) {
                    const index = thenKeyword.offset;
                    const previousChar = text.charAt(index - 1);
                    if (previousChar !== ' ') {
                        accept('error', `THEN附近需要有空格`, {
                            node: ifStatement,
                            property: 'elsifConditions'
                        });
                    }
                }
            });
        }
```

**Reply by msujew** [reply]
You can just do it like this:

```ts
  if (ifStatement) {
            const text = (AstUtils.getDocument(ifStatement).parseResult.value.$cstNode as RootCstNode).fullText;
            const thenKeywordArr = GrammarUtils.findNodesForKeyword(ifStatement.$cstNode, 'THEN');
            thenKeywordArr.forEach((thenKeyword, i) => {
                    const index = thenKeyword.offset;
                    const previousChar = text.charAt(index - 1);
                    if (previousChar !== ' ') {
                        accept('error', `THEN附近需要有空格`, {
                            node: ifStatement,
                            property: 'elsifConditions',
                            index: i
                        });
                }
            });
        }
```

**Reply by zrwsmd** [reply]
thanks a lot ,it works
